internal/pubsub/subscriber.go

```diff
@@ -0,0 +1,372 @@
+	// Attempt to create a valid resource from the URN string. If this fails, reject the message
+	resource, err := c.newResourceFromString(payload.SubjectURN)
+	if err != nil {
+		c.logger.Warnw("error parsing subject URN - will not reprocess", "event_type", eventTypeCreate, "error", err.Error())
+		return msg.Ack()
+	}
+
+	return c.createRelationships(ctx, msg, resource, payload.SubjectFields)
+}
+
+func (c *Client) handleDeleteEvent(ctx context.Context, msg *nats.Msg, payload pubsubx.Message) error {
+	// Attempt to create a valid resource from the URN string. If this fails, reject the message
+	resource, err := c.newResourceFromString(payload.SubjectURN)
+	if err != nil {
+		c.logger.Warnw("error parsing subject URN - will not reprocess", "event_type", eventTypeDelete, "error", err.Error())
+		return msg.Ack()
+	}
+
+	return c.deleteRelationships(ctx, msg, resource)
+}
+
+func (c *Client) handleUpdateEvent(ctx context.Context, msg *nats.Msg, payload pubsubx.Message) error {
+	// Attempt to create a valid resource from the URN string. If this fails, reject the message
+	resource, err := c.newResourceFromString(payload.SubjectURN)
+	if err != nil {
+		c.logger.Warnw("error parsing subject URN - will not reprocess", "event_type", eventTypeUpdate, "error", err.Error())
+		return msg.Ack()
+	}
+
+	err = c.deleteRelationships(ctx, msg, resource)
+	if err != nil {
+		return err
+	}
+
+	return c.createRelationships(ctx, msg, resource, payload.SubjectFields)
+}
+
+func (c *Client) handleUnknownEvent(ctx context.Context, msg *nats.Msg, payload pubsubx.Message) error {
+	c.logger.Warnw("unknown event - will not reprocess", "event_type", payload.EventType)
+
+	err := msg.Ack()
+	if err != nil {
+		return err
+	}
+
+	return errUnknownEvent
+}
+
+func (c *Client) receiveMsg(msg *nats.Msg) {
+	ctx, span := tracer.Start(context.Background(), "pubsub.receive", trace.WithAttributes(attribute.String("pubsub.subject", msg.Subject)))
+	defer span.End()
+
+	var payload pubsubx.Message
+
+	err := json.Unmarshal(msg.Data, &payload)
+	if err != nil {
+		c.ackWithError(msg, err)
+
+		return
+	}
+
+	resourceURN, err := urnx.Parse(payload.SubjectURN)
+	if err != nil {
+		c.ackWithError(msg, err)
+
+		return
+	}
+
+	resource, err := c.qe.NewResourceFromURN(resourceURN)
+	if err != nil {
+		c.ackWithError(msg, err)
+
+		return
+	}
+
+	c.logger.Infow("received message", "resource_type", resource.Type, "resource_id", resource.ID.String(), "event", payload.EventType)
+
+	switch payload.EventType {
+	case eventTypeCreate:
+		err = c.handleCreateEvent(ctx, msg, payload)
+	case eventTypeUpdate:
+		err = c.handleUpdateEvent(ctx, msg, payload)
+	case eventTypeDelete:
+		err = c.handleDeleteEvent(ctx, msg, payload)
+	default:
+		err = c.handleUnknownEvent(ctx, msg, payload)
+	}
+
+	if err != nil {
+		c.logger.Errorw("error handling message", "error", err.Error())
+		return
+	}
+
+	err = msg.Ack()
```

Comment:
another style comment - i prefer to see shadowed `err` with inline case statements over `err` reassignment ie.

```go
if err := msg.Ack(); err != nil {
...
}
```

☝️ imo its easier to forget a return or other thing on a reassigned error and have bugs that are hard to find bc the var is defined

although I _will_ reassign error if im initializing another var. ie

```golang

foo, err := doThing()
if err != nil {
...
}

bar, err := doOtherThing()
if err != nil {
...
}